kotest-tests/kotest-tests-core/src/jvmTest/kotlin/com/sksamuel/kotest/core/runtime/parallel/SingleInstanceTest.kt

```diff
@@ -0,0 +1,291 @@
+package com.sksamuel.kotest.core.runtime.parallel
+
+import com.sksamuel.kotest.core.runtime.PersistentThreadLocal
+import io.kotest.assertions.assertSoftly
+import io.kotest.assertions.throwables.shouldThrow
+import io.kotest.core.spec.IsolationMode
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.matchers.ints.shouldBeExactly
+import io.kotest.matchers.ints.shouldBeGreaterThan
+import io.kotest.matchers.maps.shouldHaveSize
+import io.kotest.matchers.shouldBe
+import kotlinx.coroutines.delay
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.getOrSet
+
+class SpecThreadSingleInstanceTest : FunSpec({
+
+   isolation = IsolationMode.SingleInstance
+   threadsForSpec = 3
+
+   val multipleThreadMultipleInvocationCounter =
+      PersistentThreadLocal<Int>()
+
+   afterSpec {
+      multipleThreadMultipleInvocationCounter.map.shouldHaveSize(3)
+      multipleThreadMultipleInvocationCounter.map.values.sum() shouldBe 3
+   }
+
+   test("test 1 should create own key in map with value 1") {
+      val counter = multipleThreadMultipleInvocationCounter.getOrSet { 0 }
+      multipleThreadMultipleInvocationCounter.set(counter + 1)
+   }
+
+   test("test 2 should create own key in map with value 1") {
+      val counter = multipleThreadMultipleInvocationCounter.getOrSet { 0 }
+      multipleThreadMultipleInvocationCounter.set(counter + 1)
+   }
+
+   test("test 3 should create own key in map with value 1") {
+      val counter = multipleThreadMultipleInvocationCounter.getOrSet { 0 }
+      multipleThreadMultipleInvocationCounter.set(counter + 1)
+   }
+
+   val lock = ReentrantLock()
+
+   test("test should lock object") {
+      lock.lock()
+      delay(1000)
+      lock.unlock()
+   }
+
+   test("lock should be locked") {
+      delay(300)
+      lock.isLocked shouldBe true
+   }
+
+   test("lock should be locked too") {
+      delay(300)
+      shouldThrow<AssertionError> {
+         lock.isLocked shouldBe false
+      }
+   }
+})
+
+class SpecThreadWithNestedTestSingleInstanceTest : FunSpec({
+
+   isolation = IsolationMode.SingleInstance
+   threadsForSpec = 3
+
+   val multipleThreadCounter =
+      PersistentThreadLocal<Int>()
+
+   afterSpec {
+      assertSoftly {
+         multipleThreadCounter.map.shouldHaveSize(2)
```

Comment:
We have three threads so why wouldn't this be 3. Are you expecting that only root level test are parallelized ?

Reply:
`Are you expecting that only root level test are parallelized`
Exactly. We have only two root tests:
`context("First single thread context")`
`context("Second single thread context")`
so there is only two threads will be exploited.
About root test we discussed [here](https://github.com/kotest/kotest/issues/1362#issuecomment-612313721)